File: Data Structures & Algorithms/implement-prefix-tree/submission-3.java
class TrieNode {
    public HashMap<String, TrieNode> children = new HashMap<String, TrieNode>();
    public boolean endOfWord = false;
}

class PrefixTree {

    public TrieNode root;

    public PrefixTree() {
        this.root = new TrieNode();
    }

    public void insert(String word) {
        TrieNode curr = this.root;
        for (char c: word.toCharArray()) {
            String s = "" + c;
            if (!curr.children.containsKey(s)) {
                curr.children.put(s, new TrieNode());
            }
            curr = curr.children.get(s);
        }
        curr.endOfWord = true;
    }

    public boolean search(String word) {
        TrieNode curr = this.root;
        for (char c: word.toCharArray()) {
            String s = "" + c;
            if (!curr.children.containsKey(s)) {
                return false;
            }
            curr = curr.children.get(s);
        }
        return curr.endOfWord;
    }

    public boolean startsWith(String prefix) {
        TrieNode curr = this.root;
        for (char c: prefix.toCharArray()) {
            String s = "" + c;
            if (!curr.children.containsKey(s)) {
                return false;
            }
            curr = curr.children.get(s);
        }
        return true;
    }
}

File: Data Structures & Algorithms/design-word-search-data-structure/submission-3.py
class TrieNode:

    def __init__(self):
        self.children = {}
        self.endOfWord = False

class WordDictionary:

    def __init__(self):
        self.root = TrieNode()

    def addWord(self, word: str) -> None:
        curr = self.root
        for c in word:
            if not c in curr.children:
                curr.children[c] = TrieNode()
            curr = curr.children[c]
        curr.endOfWord = True

    def search(self, word: str) -> bool:

        def dfs(node, word):
            for i in range(len(word)):
                c = word[i]
                if c == ".":
                    isWord = False
                    for key in node.children:
                        isWord = dfs(node.children[key], "" if i == len(word)-1 else word[i+1:]) or isWord
                    return isWord

                if c not in node.children:
                    return False
                else:
                    node = node.children[c]
            return node.endOfWord

        return dfs(self.root, word)
        
